yarn-project/foundation/src/json-rpc/client/safe_json_rpc_client.ts

```diff
@@ -58,10 +165,44 @@ export function createSafeJsonRpcClient<T extends object>(
     return (schema as ApiSchema)[methodName].returnType().parseAsync(response.result);
   };
 
-  const proxy: any = {};
+  const clientId = nextClientId++;
+  const proxy: any = { [CLIENT_ID]: clientId, [SEND_BATCH]: sendBatch };
   for (const method of Object.keys(schema)) {
-    proxy[method] = (...params: any[]) => request(method, params);
+    // attach the clientId to the promise so that if we want to trigger a batch immediately, we can do that
+    proxy[method] = (...params: any[]) => Object.assign(request(method, params), { [CLIENT_ID]: clientId });
   }
 
+  clients.set(clientId, new WeakRef(proxy));
```

Comment:
Instead of keeping this `clients` map, why not just attach a reference to `sendBatch` directly to the promise, and we just call that?